### PR TITLE
feat: add a gradient accent fill recipe based on hue shifting

### PR DESCRIPTION
**Pull request by bheston**
# Description

Added a gradient fill recipe based on accent color. Works by shifting the hue a currently fixed amount.

image

## Motivation & context

For the new FAST site.

Will need another version or modifier for less saturated version. Looking into an algorithm for that based on designs.

## Issue type checklist

- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

## Process & policy checklist

- [ ] I have added tests for my changes.
- [ ] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [ ] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast-dna/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

### CODE REVIEW

File: packages/react/fast-components-styles-msft/src/accent-button/index.ts [hunk 1]
```diff
@@ -24,14 +24,16 @@ import {
...
 
 const styles: ComponentStyles<AccentButtonClassNameContract, DesignSystem> = {
     ...baseButton,
     button: {
         ...buttonStyles(),
         color: accentForegroundCut,
         fill: accentForegroundCut,
-        background: accentFillRest,
+        // background: accentFillRest,
+        background: accentFillGradientRest,
```
Comment (chrisdholt) [inline packages/react/fast-components-styles-msft/src/accent-button/index.ts:36]: Do we want to actually implement this and export with this package? My assumption is no as it would ship to those leveraging the packages. If the intent here was to validate, can we revert this change?

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+>(SwatchFamilyType.rest, accentFill);
+export const accentFillGradientHover: SwatchResolver = gradientSwatchRecipe<
+    FillSwatchFamily
+>(SwatchFamilyType.hover, accentFill);
+export const accentFillGradientActive: SwatchResolver = gradientSwatchRecipe<
+    FillSwatchFamily
+>(SwatchFamilyType.active, accentFill);
+export const accentFillGradientFocus: SwatchResolver = gradientSwatchRecipe<
+    FillSwatchFamily
+>(SwatchFamilyType.focus, accentFill);
```
Comment (chrisdholt) [inline packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:55]: Should we add these to the fast-components design system provider? If this is for the site, we'll want these recipes to be reflected there so we can leverage them.

**Comment by nicholasrice** [comment]
I'm concerned that this recipe will not work with any other recipe and breaks the assumptions of the rest of the color system. How can any other recipe use the product of this algorithmically? How is this this gradient configured? How is a text color calculate for this recipe?

I'd prefer to see this private for the time being if it needs to be utilized today, but I think if gradients is a requirement then a broader look at how we make the entire system accommodate that would be prudent.

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import { DesignSystem } from "src";
```
Comment (janechu) [inline packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:1]: This pathing is going to break on the distributed package.

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    FillSwatchFamily,
+    Swatch,
+    SwatchFamily,
+    SwatchFamilyResolver,
+    swatchFamilyToSwatchRecipeFactory,
+    SwatchFamilyType,
+    SwatchRecipe,
+    SwatchResolver,
+} from "./common";
+import { accentFill } from ".";
```
Comment (janechu) [inline packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:14]: Can you make this "./index"? There's been some issue with potential path resolution between different OS's

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        const baseRecipe: SwatchRecipe = swatchFamilyToSwatchRecipeFactory<T>(
+            type,
+            callback
+        );
+        const color: Swatch = baseRecipe(designSystem);
+        const colorHsl: ColorHSL = rgbToHSL(parseColor(color));
+        let hShifted: number = colorHsl.h + 28;
+        if (hShifted > 360) {
+            hShifted -= 360;
+        }
```
Comment (janechu) [inline packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:30]: Can this be:
```
const maxColorH: number = colorHsl.h > 332
    ? colorHsl.h - 332
    : colorHsl.h + 28;
```

I think (if I understand this correctly) that we should treat the linear gradient as having a range and that you are defining the `min` and `max` values of that range so the naming should follow that logic.

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        const secondaryHsl: ColorHSL = ColorHSL.fromObject({
```
Comment (janechu) [inline packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:31]: So here instead of `secondaryHsl` maybe `maxHSL`. Since it's also used only once you could either avoid creating the const or do:
```
const maxColor: Swatch = hslToRGB(
    ColorHSL.fromObject({
        h: hShifted,
        s: colorHsl.s,
        l: colorHsl.l,
    })
).toStringHexRGB()
```

This way below you can do:
```
return format(
    "linear-gradient(0deg, {0} 0%, {1} 100%)",
    (a: any): string => minColor,
    (a: any): string => maxColor
)(designSystem);
```

I might have some of these return types wrong.

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    type: keyof T,
+    callback: SwatchFamilyResolver<T>
+): SwatchResolver {
+    return (designSystem: DesignSystem): Swatch => {
+        const baseRecipe: SwatchRecipe = swatchFamilyToSwatchRecipeFactory<T>(
+            type,
+            callback
+        );
+        const color: Swatch = baseRecipe(designSystem);
+        const colorHsl: ColorHSL = rgbToHSL(parseColor(color));
```
Comment (janechu) [inline packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:26]: Could this be `minColorHSL` instead of `colorHSL`? (see my note below)

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+function gradientSwatchRecipe<T extends SwatchFamily>(
+    type: keyof T,
+    callback: SwatchFamilyResolver<T>
+): SwatchResolver {
+    return (designSystem: DesignSystem): Swatch => {
+        const baseRecipe: SwatchRecipe = swatchFamilyToSwatchRecipeFactory<T>(
+            type,
+            callback
+        );
+        const color: Swatch = baseRecipe(designSystem);
```
Comment (janechu) [inline packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:25]: Could this be `minColor` instead of `color`? (see my note below)

File: packages/react/fast-components-styles-msft/src/accent-button/index.ts (hunk 1 above)
Comment (bheston) [reply packages/react/fast-components-styles-msft/src/accent-button/index.ts:36]: Yeah, this was mostly for validation. It could definitely go into another package more specifically for the site or FAST Frame.

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts (hunk 2 above)
Comment (bheston) [reply packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:55]: The design system provider is for the properties used to generate the recipes. In this case they build from the base accent recipe. We could potentially add properties for the hue offset and gradient parameters like start / end % and angle.

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts (hunk 5 above)
Comment (bheston) [reply packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:30]: 360 is significant because it's the hue range for colors. 28 is the variable here, and 332 is just 360 - 28. I'd actually prefer this logic be in the color package, as hue is often thought of as a repeating circle, especially for wrapping like this. We could update this here, but 332 is not significant, so I wouldn't want to hard-code that. Actually, 28 could possibly be a design system property if building this for wider use.

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts (hunk 6 above)
Comment (bheston) [reply packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:31]: I took this mostly. I went with 'shifted' instead of 'max' though. I don't think it's as much min / max as it is original / shifted (or something along those lines). For instance, if the shift were -28 instead of +28, what would be min / max at that point?

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts (hunk 8 above)
Comment (bheston) [reply packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:25]: Made it baseColor.

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts (hunk 7 above)
Comment (bheston) [reply packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:26]: Made it baseColorHSL.

**Comment by bheston** [comment]
@nicholasrice your points occurred to me as I was working on this. I'm not sure the right way to calculate the contrast over a gradient, if it's possible to use the strongest difference between A and B, or if it's possible the colors in between could be less contrast than A or B. In the current usage it's not being used for the color calculation, but letting the base accent color stand in. As you point out, that may not be safe, and not the way the recipes are intended to chain. To fully support gradients (or perhaps acrylic) I think we'll need to extend the system. The color 'value' is going to have to be provided differently than the actual style value.

File: packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts (hunk 6 above)
Comment (janechu) [reply packages/react/fast-components-styles-msft/src/utilities/color/accent-fill-gradient.ts:31]: Good point

**Comment by bheston** [comment]
This will likely live somewhere else, once we address how to handle it within the recipe infrastructure and for a11y.
